lib/spack/spack/cmd/bootstrap.py

```diff
@@ -249,6 +265,57 @@ def _status(args):
         print()
 
 
+def _add(args):
+    initial_sources = spack.bootstrap.bootstrapping_sources()
+    names = [s['name'] for s in initial_sources]
+
+    # If the name is already used error out
+    if args.name in names:
+        msg = 'a source named "{0}" already exist. Please choose a different name'
+        raise RuntimeError(msg.format(args.name))
+
+    # Check that the metadata file exists
+    file = spack.util.path.canonicalize_path(args.metadata)
+    if not os.path.exists(file):
+        msg = 'the file "{0}" does not exist'
+        raise RuntimeError(msg.format(args.metadata))
+
+    # Insert the new source as the highest priority one
+    write_scope = args.scope or spack.config.default_modify_scope(section='bootstrap')
+    sources = spack.config.get('bootstrap:sources', scope=write_scope) or []
+    sources = [
+        {'name': args.name, 'metadata': args.metadata}
+    ] + sources
+    spack.config.set('bootstrap:sources', sources, scope=write_scope)
+
+    msg = 'New bootstrapping source "{0}" added in the "{1}" configuration scope'
+    llnl.util.tty.msg(msg.format(args.name, write_scope))
+
+
+def _remove(args):
+    initial_sources = spack.bootstrap.bootstrapping_sources()
+    names = [s['name'] for s in initial_sources]
+    if args.name not in names:
+        msg = ('cannot find any bootstrapping source named "{0}". '
+               'Run `spack bootstrap list` to see available sources.')
+        raise RuntimeError(msg.format(args.name))
+
+    for current_scope in spack.config.scopes():
+        sources = spack.config.get('bootstrap:sources', scope=current_scope) or []
+        if args.name in [s['name'] for s in sources]:
+            sources = [s for s in sources if s['name'] != args.name]
+            spack.config.set('bootstrap:sources', sources, scope=current_scope)
+            msg = ('Removed the bootstrapping source named "{0}" from the '
+                   '"{1}" configuration scope.')
+            llnl.util.tty.msg(msg.format(args.name, current_scope))
+        trusted = spack.config.get('bootstrap:trusted', scope=current_scope) or []
+        if args.name in trusted:
+            trusted.pop(args.name)
+            spack.config.set('bootstrap:trusted', trusted, scope=current_scope)
+            msg = 'Deleting information on "{0}" from list of trusted sources'
+            llnl.util.tty.msg(msg.format(args.name))
```

Comment:
I think we can restructure this to avoid some unnecessary reading from the config system

```suggestion
def _remove(args):
    for current_scope in spack.config.scopes():
        sources = spack.config.get('bootstrap:sources', scope=current_scope) or []
        if args.name in [s['name'] for s in sources]:
            sources = [s for s in sources if s['name'] != args.name]
            spack.config.set('bootstrap:sources', sources, scope=current_scope)
            msg = ('Removed the bootstrapping source named "{0}" from the '
                   '"{1}" configuration scope.')
            llnl.util.tty.msg(msg.format(args.name, current_scope))
        trusted = spack.config.get('bootstrap:trusted', scope=current_scope) or []
        if args.name in trusted:
            trusted.pop(args.name)
            spack.config.set('bootstrap:trusted', trusted, scope=current_scope)
            msg = 'Deleting information on "{0}" from list of trusted sources'
            llnl.util.tty.msg(msg.format(args.name))
            return

    # error if we did not find a source of this name
    msg = ('cannot find any bootstrapping source named "{0}". '
           'Run `spack bootstrap list` to see available sources.')
    raise RuntimeError(msg.format(args.name))
```

Reply:
This is the only comment where I slightly disagree. I don't think there will be major improvements in performance by deferring error handling as the last thing (due to both this operation being not "critical" and configuration file reads being memoized). 

On the other hand, I think sanitizing input as the first operation and then dealing only with the "happy" path makes function usually more readable.